### ISSUE TITLE
Implement spinloop detection for ARM exclusive pairs

### ISSUE THREAD

**Issue by natgavrilenko** [hernanponcedeleon/Dat3M#719]
Consider the attached 'spinloop.litmus' test. The loop in thread P0 has no side effects, because this loop terminates whenever STLXR is executed. The value of W0 is 0 if the store was successful and 1 if the store failed. However, the test terminates with UNKNOWN result for any unrolling bound. The manually unrolled test 'spinloop-unrolled.litmus' works as expected.

The current implementation of spinloop detection fails to detect this spinloop. The reason is that dartagnan internally represent STLXR as a pair of instructions in the same branch, while spinloop detection essentially analyses the branches withing a loop. 

Note: I would avoid changing the internal representation of ARM exclusive pair to contain branches. Representing such event pairs using branches would generate artificial control dependencies, which are used in the ARM consistency model. I would rather implement a special handling of RmwStoreExclusive events in the spinloop detection method.

[spinloop.txt](https://github.com/user-attachments/files/16695825/spinloop.txt)
[spinloop-unrolled.txt](https://github.com/user-attachments/files/16695826/spinloop-unrolled.txt)

**Comment by ThomasHaas** [comment]
Your code actually shows a problem in the parser for AARCH64. The registers W0 is parsed as X0, making the loop truly side-effectful as the address of the exclusive load (X0) gets (wrongly) updated by the status of the exclusive store (W0).
That being said, we would still fail to detect the side-effectfree iteration where the store fails (but fixing this is easy, though insufficient due to the parser bug).

**Comment by hernanponcedeleon** [comment]
Register W0 is the lowest 32 bits of register X0 (which is 64 bits). The fact that the exclusive store partially updates the address of the load looks to me like a problem in the test per se.

I think @db7 generated this program from some real assembly. Diogo can you confirm if the assembly really uses W0/X0 for both the addresses and the execution status of the exclusive store, or this was a problem when translating from assembly to pasm with kittens?

**Comment by ThomasHaas** [comment]
Oh, I'm not familiar with the register names in assembly code, but that certainly sounds wrong.

**Comment by db7** [comment]
@hernanponcedeleon I can check that.

**Comment by ThomasHaas** [comment]
Yes, but W0 is the lower part of X0, so the address gets updated if W0 is changed. This is surely wrong.

**Comment by hernanponcedeleon** [comment]
```
 LC00:
 LDAXR   X3, [X0]
 STLXR   W0, X2, [X0]
 CBZ     W0, LC01
 B       LC00
```
In the first iteration, the load accesses address `X0` (which contains the address of some variable `x`). If the exclusive store fails, `W0` gets value 1, meaning that the lowest 32 bits of `X0` were updated. Now in the second iteration the address being accessed by the load was changed (unless the lowest 32 bits were already 1, but this does not sound like a reasonable assumption).

**Comment by hernanponcedeleon** [comment]
FYI, herd7 complains about this test with `Warning: File "arm.litmus": Non-symbolic memory access found on '[1]' (User error)` but it works if I replace the register used for the status (`W0`) to something else like `W8`.

@db7 is this code generated by a compiler or using inline assembly?

**Comment by ThomasHaas** [comment]
It does not only update the lowest bit but writes the whole value. So it should be wrong almost 100% of the time.

**Comment by db7** [comment]
Got it. It sounds fishy.

**Comment by ThomasHaas** [comment]
Definitely looks like a translation error. Nevertheless, the spinloop issue remains though I have a fix ready. I just was confused when the fix didn't work and then I looked into the code, finding this odd usage of registers.

**Comment by db7** [comment]
Ok, so you confirm the issue, just the reported example was itself buggy :-D

Please let me know if you have a branch with the code pushed. I can try it out in several tests on my side.

**Comment by ThomasHaas** [comment]
Ok, there is another issue. If I add proper detection, then such loops are detected as liveness violations because the exclusive store can fail spuriously forever. I guess one needs to argue about fairness again and assume that the store will eventually succeed.

**Comment by db7** [comment]
How do you handle progress of `atomic_exchange` at the C level?

**Comment by ThomasHaas** [comment]
We don't implement those with a store that can fail and thus do not even require a loop.
I'm not sure if we correctly handle weak CAS though (we might always upgrade it to a strong one?).

**Comment by hernanponcedeleon** [comment]
> How do you handle progress of `atomic_exchange` at the C level?

At the C level, the exchange cannot fail. The issue is that in arm, ["An implementation might clear an exclusive monitor between the LDREX and the STREX, without any application-related cause."](https://developer.arm.com/documentation/ddi0406/c/Application-Level-Architecture/Application-Level-Memory-Model/Synchronization-and-semaphores/Load-Exclusive-and-Store-Exclusive-usage-restrictions), making the store to "spuriously" fail.

**Comment by ThomasHaas** [comment]
For now you can't. We could add an option to make all exclusive stores succeed (dangerous if used outside loops) or add the fairness assumption for the purpose of liveness detection.

**Comment by hernanponcedeleon** [comment]
> EDIT: The fairness assumption is actually valid: the specification says that forward progress should be guaranteed.

It should be enough to make the spinloop instrumentation pass to add an `assume(status_reg)`, right?

**Comment by ThomasHaas** [comment]
Yes, but that is kinda cheating. If we do an instrumentation like this, I would rather have a dedicated pass. Either way, the pass would likely be unsound in general or be specialized to only work on specific shapes of loops (i.e., be missing more exotic usages of exclusive stores).

**Comment by ThomasHaas** [comment]
> Ok, so fairness assumption is default. I don't have to configure anything?

I opened a new PR with the necessary changes (#722). You can try if it solves your issues.

**Comment by hernanponcedeleon** [comment]
@db7 can you check if with these changes we now return PASS in all the tests from kittens?

**Comment by db7** [comment]
The branch does not accept lowercase registers. Herd7 does.
I think I was working on a branch that would accept lowercase registers.
Are you planning to change that or should I always use upper case registers?

**Comment by db7** [comment]
OK, I changed that on my side, but you might want to review the requirement of uppercase of register names in the future.

**Comment by db7** [comment]
This [test.txt](https://github.com/user-attachments/files/16816061/test.txt) file is failing with some weird conditional thing. I didnt have that before.

**Comment by db7** [comment]
Wait, the file I sent before to you was working with Datargnan before and the file was using lower case registers. I mean this one:

```
AArch64 test.litc
{
  0:x0 = y;
  0:x1 = x;
  0:x2 = r0_P0;
  1:x0 = y;
  1:x1 = x;
  1:x2 = r0_P1;
}

P0                         | P1                     ;
                           |                        ;
      MOV     w3, #2       |       MOV     w3, #1   ;
                           |                        ;
LC78:                      |       LDR     w0, [x0] ;
      LDAXR   w5, [x1]     |                        ;
                           |       STLR    w3, [x1] ;
      STLXR   w4, w3, [x1] |                        ;
                           |       STR     w0, [x2] ;
      CBNZ    w4, LC78     |                        ;
                           |                        ;
      MOV     w3, #1       |                        ;
                           |                        ;
      STR     w3, [x0]     |                        ;
                           |                        ;
      STR     w5, [x2]     |                        ;

exists (x = 2 /\ (r0_P0 = 1 /\ r0_P1 = 1))

**Comment by ThomasHaas** [comment]
This error is not new to this PR. It happens because we have not enabled our loop normalization pass on Litmus code (we are somewhat more careful on transforming litmus code). I will add a PR that enables this pass.

**Comment by hernanponcedeleon** [comment]
This is already enabled I the `arm-asm` branch. @db7 can you apply the changes of this PR on top of that branch and confirm it works as expected?

**Comment by ThomasHaas** [comment]
I guess we should revise/improve our handling of litmus code, seeing that it sees more usage nowadays.

**Comment by db7** [comment]
@hernanponcedeleon, thank you! I was using `arm-asm` for a long time and couldnt even remember.

**Comment by ThomasHaas** [comment]
Can this be closed?

**Comment by hernanponcedeleon** [comment]
Fixed by 6326445a3923987549c4299678a6fc5cc3d4c9a8
